### PR TITLE
Vagrant: Upgrade to Fedora 21, Docker 1.6, clean-up SDN

### LINKED ISSUE #8834: Vagrant/Virtualbox - Minion Kublet fails to connect to master
The minion-1 kublet is once again failing to connect to the master server.  Previously the problem was the default port change. This time it's less clear to me.

````
[mlamouri@blade16 kubernetes]$ kubectl cluster-info
Kubernetes master is running at https://10.245.1.2
KubeDNS is running at https://10.245.1.2/api/v1beta3/proxy/namespaces/default/services/kube-dns
````

_minon-1 kubelet config_
````
[vagrant@kubernetes-minion-1 ~]$ cat /etc/sysconfig/kubelet 
# TODO: remove nginx for other cloud providers.
# Disable registration for the kubelet running on the master on GCE.
# TODO(roberthbailey): Make this configurable via an env var in config-default.sh



  
  


DAEMON_ARGS=" --api_servers=https://10.245.1.2:6443  --hostname_override=10.245.1.3 --cloud_provider=vagrant --config=/etc/kubernetes/manifests --allow_privileged=False --v=2 --cluster_dns=10.247.0.10 --cluster_domain=cluster.local  "
````

_minion-1 kube-proxy config_
````
[vagrant@kubernetes-minion-1 ~]$ cat /etc/sysconfig/kube-proxy 


# TODO: remove nginx for other cloud providers.
  DAEMON_ARGS=" --master=https://10.245.1.2:6443 --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2"
````

_kubelet status_
````
[minion-1] sudo systemctl status kubelet -l
kubelet.service - Kubernetes Kubelet Server
   Loaded: loaded (/usr/lib/systemd/system/kubelet.service; enabled)
   Active: active (running) since Tue 2015-05-26 20:22:24 UTC; 18min ago
     Docs: https://github.com/GoogleCloudPlatform/kubernetes
 Main PID: 2983 (kubelet)
   CGroup: /system.slice/kubelet.service
           ├─2983 /usr/local/bin/kubelet --api_servers=https://10.245.1.2:6443 --hostname_override=10.245.1.3 --cloud_provider=vagrant --config=/etc/kubernetes/manifests --allow_privileged=False --v=2 --cluster_dns=10.247.0.10 --cluster_domain=cluster.local
           └─3018 journalctl -f

May 26 20:40:38 kubernetes-minion-1 kubelet[2983]: E0526 20:40:38.984396    2983 kubelet.go:728] Unable to construct api.Node object for kubelet: failed to get instance ID from cloud provider: Post http://127.0.0.1:8000/login: dial tcp 127.0.0.1:8000: connection refused
May 26 20:40:45 kubernetes-minion-1 kubelet[2983]: E0526 20:40:45.986972    2983 kubelet.go:728] Unable to construct api.Node object for kubelet: failed to get instance ID from cloud provider: Post http://127.0.0.1:8000/login: dial tcp 127.0.0.1:8000: connection refused
May 26 20:40:46 kubernetes-minion-1 kubelet[2983]: E0526 20:40:46.150296    2983 file.go:53] Unable to read config path "/etc/kubernetes/manifests": path does not exist, ignoring
May 26 20:40:48 kubernetes-minion-1 kubelet[2983]: E0526 20:40:48.476591    2983 kubelet.go:1518] error getting node: node 10.245.1.3 not found
May 26 20:40:52 kubernetes-minion-1 kubelet[2983]: E0526 20:40:52.988425    2983 kubelet.go:728] Unable to construct api.Node object for kubelet: failed to get instance ID from cloud provider: Post http://127.0.0.1:8000/login: dial tcp 127.0.0.1:8000: connection refused
May 26 20:40:58 kubernetes-minion-1 kubelet[2983]: E0526 20:40:58.480547    2983 kubelet.go:1518] error getting node: node 10.245.1.3 not found
May 26 20:40:59 kubernetes-minion-1 kubelet[2983]: E0526 20:40:59.996271    2983 kubelet.go:728] Unable to construct api.Node object for kubelet: failed to get instance ID from cloud provider: Post http://127.0.0.1:8000/login: dial tcp 127.0.0.1:8000: connection refused
May 26 20:41:06 kubernetes-minion-1 kubelet[2983]: E0526 20:41:06.199604    2983 file.go:53] Unable to read config path "/etc/kubernetes/manifests": path does not exist, ignoring
May 26 20:41:06 kubernetes-minion-1 kubelet[2983]: E0526 20:41:06.997534    2983 kubelet.go:728] Unable to construct api.Node object for kubelet: failed to get instance ID from cloud provider: Post http://127.0.0.1:8000/login: dial tcp 127.0.0.1:8000: connection refused
May 26 20:41:08 kubernetes-minion-1 kubelet[2983]: E0526 20:41:08.485200    2983 kubelet.go:1518] error getting node: node 10.245.1.3 not found
````

### LINKED ISSUE #8907: Vagrant provision script - avoiding environment settings
In the provision script it tries to do `grep -v` which is an inverted match for `/^#...` regex expression. On some systems users _may_ have colors set via the user's profile. The result is a concat with color codes injected into the resulting provision scripts, ultimately failing the provision and failing `vagrant up` command.

It may be better to avoid this and other edge cases by using awk.

I have a fork with this change:
https://github.com/JasonGiedymin/kubernetes/commit/f25ecb7e4689d751a66ffcf1bab42ae56bcadc4e

Please review.

### PR DESCRIPTION
**Pull request by derekwaynecarr**
PR makes the following changes:
1. Move to Fedora 21 since Fedora 20 no longer packages latest docker updates
2. Make Salt configure an explicit Docker environment file (/etc/sysconfig/docker) on RedHat based env
3. Disable registration of kubelet on master just like GCE for Vagrant mode
4. Move SDN configuration out of Salt and as part of basic provision script to improve reliability, and enable us to manage the docker defaults option globally rather than the hack we had before.

Fixes #8907
Fixes #8834

### CODE REVIEW

**Comment by JasonGiedymin** [comment]
phew, that is a big pr...

File: cluster/vagrant/util.sh [hunk 1]
```diff
@@ -161,8 +163,8 @@ function create-provision-scripts {
       echo "CONTAINER_SUBNET='${CONTAINER_SUBNET}'"
       echo "DOCKER_OPTS='${EXTRA_DOCKER_OPTS-}'"
       echo "VAGRANT_DEFAULT_PROVIDER='${VAGRANT_DEFAULT_PROVIDER:-}'"
-      grep -v "^#" "${KUBE_ROOT}/cluster/vagrant/provision-minion.sh"
-      grep -v "^#" "${KUBE_ROOT}/cluster/vagrant/provision-network.sh"
+      awk '!/^#/' "${KUBE_ROOT}/cluster/vagrant/provision-network.sh"
```
Comment (skonzem) [inline cluster/vagrant/util.sh]: Looks like some extra whitespace slipped in.

**Comment by bgrant0607** [comment]
@derekwaynecarr Any suggestions re. who would be a good reviewer for this?

**Comment by derekwaynecarr** [comment]
@bgrant0607 - is it ok if someone from our side tests and comments success?  typically i am the reviewer for this provider, but I cannot really self-review.
 
@csrwng - would you mind testing that you can bring up a vagrant cluster from this branch and commenting upon your hopeful success?

```
$ make clean
$ make release-skip-tests
$ cluster/kube-up.sh
$ cluster/kubectl.sh get pods
```

**Comment by csrwng** [comment]
@derekwaynecarr - vagrant succeeded for me on this branch. Everything's looking good.

File: cluster/vagrant/util.sh (hunk 1 above)
Comment (bgrant0607) [inline cluster/vagrant/util.sh]: ping

**Comment by bgrant0607** [comment]
Thanks, @csrwng.

File: cluster/vagrant/util.sh (hunk 1 above)
Comment (derekwaynecarr) [inline cluster/vagrant/util.sh]: Away from computer now, will fix white space tomorrow am

**Comment by bgrant0607** [comment]
needs rebase

**Comment by derekwaynecarr** [comment]
Rebased, should be good to go in on-call queue.

**Comment by bgrant0607** [comment]
Rerunning Shippable.

**Comment by derekwaynecarr** [comment]
Shippable has stopped flaking on this PR, I think this is good to leave the nest.

**Comment by fahadpolash** [comment]
Hello,
I was trying to bringing up the cluster from the master brunch.
Its giving me the following errors. Is there any issues with fedora 21 repos?


==> master: It looks like the required network bridge has not yet been created
==> master: Installing, enabling prerequisites
==> master: 
==> master: 
==> master:  One of the configured repositories failed (Fedora 21 - x86_64),
==> master:  and yum doesn't have enough cached data to continue. At this point the only
==> master:  safe thing yum can do is fail. There are a few ways to work "fix" this:
==> master: 
==> master:      1. Contact the upstream for the repository and get them to fix the problem.
==> master: 
==> master:      2. Reconfigure the baseurl/etc. for the repository, to point to a working
==> master:         upstream. This is most often useful if you are using a newer
==> master:         distribution release than is supported by the repository (and the
==> master:         packages for the previous distribution release still work).
==> master: 
==> master:      3. Disable the repository, so yum won't use it by default. Yum will then
==> master:         just ignore the repository until you permanently enable it again or use
==> master:         --enablerepo for temporary usage:
==> master: 
==> master:             yum-config-manager --disable fedora
==> master: 
==> master:      4. Configure the failing repository to be skipped, if it is unavailable.
==> master:         Note that yum will try to contact the repo. when it runs most commands,
==> master:         so will have to try and fail each time (and thus. yum will be be much
==> master:         slower). If it is a very temporary problem though, this is often a nice
==> master:         compromise:
==> master: 
==> master:             yum-config-manager --save --setopt=fedora.skip_if_unavailable=true
==> master: 
==> master: Cannot retrieve metalink for repository: fedora/21/x86_64. Please verify its path and try again
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

**Comment by JasonGiedymin** [comment]
From this PR my system has not been working right. Other people have been experiencing issues as well. #9488

**Comment by karlkfi** [comment]
Fedora 21 switches from initd to systemd, but it looks like the salt scripts were not updated to work with systemd. The kube-apiserver specifically fails to start on vagrant because `/etc/sysconfig/kube-apiserver` is not being generated. See https://github.com/GoogleCloudPlatform/kubernetes/issues/9681 for my debugging comments.

**Comment by derekwaynecarr** [comment]
Fedora 20 also used systemd.  

Vagrant is fixed when this PR merges:
https://github.com/GoogleCloudPlatform/kubernetes/pull/9563

API server is run as pod by kubelet.  Same for scheduler, etcd, and controller manager.  Exact same setup as GCE.  

Only the kubelet and docker is managed by systemd now.

**Pull request merged**